---
title: "Female vs Male"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_knit$set(fig.width=12, fig.height=8, fig.path='Figs/',
                      echo=FALSE, 
                      warning=FALSE, 
                      message=FALSE,
                      root.dir="~/BTSync/FetalRNAseq/Github/GENEX-FB1/")
```

```{r, incluse=FALSE}
library(tidyverse)
library(gridExtra)
library(clusterProfiler)
library(GGally)
library(stringr)
source("Shiny/GENEX-FB1/FormatGGplot.R")

LibraryInfo <- read_tsv("Data/SampleInfo.txt", 
                        col_types=cols(Sample='c')
                       ) %>%
full_join(read_tsv("Data/experiment.txt", 
                        col_types=cols(Sample='c')
                       )
          )

gene_info <- read_tsv("Data/genes.txt") %>%
  mutate(gene_id = sub("\\.[0-9]+", "", gene_id),
         ChrType = ifelse(seqid == 'ChrX' | seqid == 'ChrY', seqid, 'autosomal')
         ) %>%
  dplyr::select(Id = gene_id, SYMBOL=gene_name, Chr=seqid, ChrType)

MalevsFemale_complete <- read_tsv("Results/MvsF_12_20_PCW_FDR_0.1/tables/MalevsFemale.complete.txt")%>%
  mutate(Id = sub("(ENSG[0-9]+)\\.[0-9]+", '\\1', Id))

MalevsFemale_complete <- right_join(gene_info, MalevsFemale_complete)

Fold_changes <- read_tsv("Shiny/GENEX-FB1/Data/fitted.txt")

CPM1 <- Fold_changes %>%
  filter(! is.na(log2FoldChange)) %>%
  group_by(ageBin) %>%
  dplyr::summarise(n=n()) %>%
  filter(ageBin != '12-19') %>%
  mutate(week=ifelse(ageBin=='15-16', 15.5,
                     ifelse(ageBin=='17-19', 18, as.numeric(ageBin))
        ))


PlotFC_by_ChrType <- function(genes, set_name) {
  genes <- filter(genes, ageBin != '12-19') %>%
    mutate(week = ifelse(ageBin=='15-16', 15.5, ifelse(ageBin == '17-19', 18, as.numeric(ageBin))))
  plot <- ggplot(genes, aes(x=week, y=log2FoldChange, group=Id)) +
    geom_line(alpha=1, aes(colour=ChrType)) +
    geom_point(colour='black', size=1, alpha=.25, data=filter(genes, padj<.1)) +
    scale_x_continuous(breaks=c(12,13,14,15.5,18), labels=c('12', '13', '14', '15-16', '17-19')) +
    scale_colour_manual(values=brewer.pal(6, "Set1")[c(1,2,4)]) + 
    main_theme() +
    ylab("<--- Female-biased        Male-biased --->") +
    xlab("Post Conception Week") +
    theme(axis.title.y=element_text(size=9)) +
    ggtitle(paste("Log fold differences in", set_name))
  plot        
}

```

## Histogram of sample size per week/sex
```{r}

LibraryInfo %>% group_by(Sex) %>% summarise(mean(PCW))
layout(matrix(c(1,2), 2, byrow = TRUE))
p1 <- ggplot(LibraryInfo, aes(x=PCW, fill=Analysis)) +
  geom_bar() +
  facet_grid(Sex ~ .) +
  main_theme() +
  scale_y_continuous(breaks=seq(0,100, 2)) +
  scale_x_continuous(breaks=seq(11,19, 1)) +
  xlab("Post Conception Weeks") +
  theme(axis.text.x=element_text(size=8)) +
  scale_fill_brewer(type = "qual", palette = 6) 

LibraryInfo <- mutate(LibraryInfo, age_bin =ifelse(LibraryInfo$PCW < 15, LibraryInfo$PCW,
                                     ifelse(LibraryInfo$PCW < 17, '15-16', '17-19')))
dplyr::select(LibraryInfo, age_bin, Sex) %>% table()
p2 <- ggplot(LibraryInfo, aes(x=age_bin, fill=Analysis )) +
  geom_bar() +
  facet_grid(Sex ~ .) +
  main_theme() +
  scale_y_continuous(breaks=seq(0,100, 2)) +
  #scale_x_continuous(breaks=c(11,12,13,14,15,16),
  #                   labels=c('11','12','13','14','15-16','17-19')) +
  xlab("Post Conception Weeks") +
  theme(axis.text.x=element_text(size=8)) +
  theme(legend.position='right') +
  scale_fill_brewer(type = "qual", palette = 6) 

grid.arrange(p1, p2, ncol=2)
```


## R Markdown

- 51 million - 353 million read pairs (14 billion total)
- 28 million - 252 million properly mapped pairs (10.7 billion total)
- 5.3 million - 114 million read pairs included in counts matrix (3.5 billion total)

- Samples with < 10 million read pairs in count matrix:
    - 17046 (12 PCW) has 78% of reads mapping to rDNA
    - 18655 (13 PCW) has relatively low sequencing depth (78 million read pairs) once I exluded the first batch of sequence because I don't know how to combine technical replicates with different batch effects. It also has a lot of ribosomal
    
- one additional samples with <10% of read pairs in count matrix has ~50% of reads mapping to rDNA (16428)

- 11971 (17 PCW) has 671 million read pairs and >450 million in count matrix


```{r }
count_matrix <- MalevsFemale_complete %>%
  dplyr::select(-Id, -SYMBOL, -Chr, -starts_with('norm'), -baseMean, -Female, -Male, -FC, -log2FoldChange, -pvalue, -padj, -tagwise.dispersion, -trended.dispersion, -Chr, -ChrType)
Counts <- count_matrix %>% 
  tidyr::gather(Sample, Count) %>% 
  mutate(Sample = as.character(Sample)) %>% 
  group_by(Sample) %>%
  dplyr::summarise(TotalCount=sum(Count))

lengths <- read_delim("FastQC/seq_lengths.txt", 
                      delim='\t', 
                      col_names=c('read_file', 'num_reads'), 
                      col_types=cols(read_file='c', num_reads='i')
                      )
SeqInfo <- read_delim("Data/sequences.txt", 
                     delim='\t',
                     col_names=c('read_file', 'read_group', 'centre', 'folder', 'trimmed'),
                     col_types=cols(read_file='c', read_group='c', centre='c', folder='c', trimmed='c')
                     )


combined <- left_join(SeqInfo, lengths, by='read_file') %>%
  separate(read_group, c('Sample', 'run'), fill='right', remove=FALSE) %>%
  filter(grepl('_1$', read_file) | grepl('_R1_', read_file))

mapping_stats <- read_tsv("Tables/read_numbers.txt",
                          col_types = cols(sample = 'c'),
                          trim_ws = TRUE) %>%
  mutate(sample = str_extract(sample, "^[^-]+")) %>% 
  group_by(sample) %>% 
  summarise_each("sum")


total_reads <- combined %>%
  group_by(Sample) %>%
  dplyr::summarise(TotalFragments=sum(as.numeric(num_reads)))

final_stats <- total_reads %>%
  full_join(mapping_stats, by=c('Sample' = 'sample')) %>%
  right_join(Counts) %>% 
  mutate(MappedFragments = Paired/2, rDNA_Fragments= floor(rDNA/2)) %>%
  dplyr::select(Sample, TotalFragments, rDNA_Fragments, MappedFragments, TotalCount) %>%
  mutate(perc_rDNA = rDNA_Fragments/TotalFragments*100,
         perc_mapped = MappedFragments/TotalFragments*100,
         perc_counted = TotalCount/TotalFragments*100)

final_stats %>% arrange(TotalCount)
final_stats %>%
  dplyr::select(TotalFragments, MappedFragments, TotalCount) %>%
  summarise(TotalFragments=sum(as.numeric(TotalFragments))/1000000000,
            MappedFragments=sum(as.numeric(MappedFragments)/1000000000),
            TotalCount=sum(as.numeric(TotalCount)/1000000000)
  )


full_join(LibraryInfo, final_stats, by='Sample') %>% 
  write_tsv("Tables/CountSummary.txt")

```

- In full 105 sample dataset, 47524 of 66023 ENSEMBL features have at least one read mapping
- 17070 features with >=1 Counts Per Million in at least as many samples as the min of male/famale were included in the differential expression analysis


```{r}
feature_stats <- data.frame(Samples=c(ncol(count_matrix)), TotalFeatures=c(nrow(count_matrix)))
feature_stats <- count_matrix %>%
  mutate(sum=rowSums(.)) %>%
  filter(sum > 0) %>%
  dplyr::summarise(NonZeroFeatures=n()) %>%
  bind_cols(feature_stats)
feature_stats <- MalevsFemale_complete %>%
  filter(! is.na(FC)) %>%
  dplyr::summarise(IncludedFeatures=n()) %>%
  bind_cols(feature_stats)
  
dplyr::select(feature_stats, Samples, TotalFeatures, NonZeroFeatures, IncludedFeatures)

p1 <- count_matrix %>%
  gather(Sample, Count) %>%
  mutate(Sample = as.character(Sample)) %>%
  filter(Count > 0) %>%
  group_by(Sample) %>%
  dplyr::summarise(n=n()) %>%
  full_join(dplyr::select(LibraryInfo, Sample, PCW, Sex)) %>% 
  ggplot(aes(x=PCW, y=n)) +
    geom_bar(aes(x=week, y=n), stat='identity', data=CPM1) +
    geom_jitter(aes(colour=Sex), height=0, width=.25) +
    main_theme() +
    scale_y_continuous(limits=c(0,66000), breaks=seq(0,60000,10000)) +
    scale_x_continuous(breaks=seq(12, 19, 1)) +
    xlab("Post Conception Weeks") +
    ylab("Ensembl transcripts (66023 total)") +
    theme(axis.text.x=element_text(size=8)) +
    scale_colour_brewer(type = "qual", palette = 6) +
    theme(legend.position="right")
p1

```


```{r}
PlotTimepoint<-function(geneID, counts, fittedBias, target, ages, title) {
  geneID = str_replace(geneID, '(ENSG\\d+)\\.\\d+', '\\1')
  ageSplit <- strsplit(ages, '-')[[1]]
  min <- ageSplit[1]
  max <- ageSplit[length(ageSplit)]
  data <- counts %>% filter(SYMBOL == geneID | Id == geneID) %>%  
    dplyr::select(-SYMBOL, -Id, -Chr) %>%
    gather() %>%
    separate(key, into=c('norm', 'Sample'), sep='[.]') %>%
    dplyr::select(Sample, value) %>%
    left_join(target) %>%
    filter(PCW >= min & PCW <= max)
  mean <- filter(fittedBias, SYMBOL == geneID | Id == geneID) %>% filter(ageBin==ages)
  fc <- mean$log2FoldDiff[1] %>% format(digits=2)
  pval <- mean$pvalue[1]
  qval <- mean$padj[1]
  mean <- mean %>% dplyr::select(Male, Female) %>%
    gather('Sex', 'mean')
  plot<-  ggplot(data, aes(x=Sex, colour=Sex)) + 
    geom_jitter(aes(y=value, shape=Analysis), height = 0, width=.1, alpha=.75) + 
    geom_errorbar(aes(ymin=mean, ymax=mean), colour='black', size=1, width=.5, data=mean) +
    ylab("Normalised counts") +
    xlab('') +
    figure_theme() +
    scale_colour_brewer(type = "qual", palette = 6) +
     theme(plot.title=element_text(face="italic")) +
    ggtitle(title) 
  plot
}

fittedBias <- read_delim("Shiny/GENEX-FB1/Data/fitted.txt", "\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::rename(log2FoldDiff = log2FoldChange) %>%
  mutate(Sex = ifelse(log2FoldDiff < 0, 'F', 'M'), 
         ChrType = ifelse(Chr %in% c('chrX','chrY'), Chr, 'autosome'), 
         Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) 

fittedBiasAll <- filter(fittedBias, ageBin == '12-19')
fittedBias <-  filter(fittedBias, ageBin != '12-19')

counts <- read_delim("Shiny/GENEX-FB1/Data/counts.txt", "\t", escape_double = FALSE, trim_ws = TRUE)

Fig3a <- PlotTimepoint('GABRA1', counts, fittedBias, LibraryInfo, '14', 'GABRA1')
Fig3b <- PlotTimepoint('GRIN2A', counts, fittedBias, LibraryInfo, '14', 'GRIN2A')
Fig3c <- PlotTimepoint('NES', counts, fittedBias, LibraryInfo, '14', 'NES')
Fig3d <- PlotTimepoint('EOMES', counts, fittedBias, LibraryInfo, '14', 'EOMES')

grid.arrange(Fig3a, Fig3b, Fig3c, Fig3d, ncol=2)
```

# Genes that are DE in combined analysis

- LINC01597 consistently 2 fold higher in males at all timepoints
- X chromosome genes consistently slightly lower in males at all timepoints
```{r}


all_DE_by_week <- filter(Fold_changes, padj < 0.1 & ageBin=='12-19') %>% 
  select(Id) %>%
  left_join(Fold_changes) %>% 
  mutate(ChrType=ifelse(Chr=='chrX', 'chrX', ifelse(Chr=='chrY', 'chrY', 'autosome')))

PlotFC_by_ChrType(all_DE_by_week, 'All DE')
all_DE_by_week %>% filter(ChrType != 'chrY' & SYMBOL != 'XIST') %>%
  PlotFC_by_ChrType("Non Y/XIST DE")
```

# Genes that are DE in any analysis

- There are a number of autosomal genes that are expressed in both sexes at week 12, only in females at week 13 and in neither sex at later timepoints
```{r}
all_DE_by_week <- filter(Fold_changes, padj < 0.1) %>% 
  select(Id) %>%
  group_by(Id) %>%
  summarise(n=n()) %>% #don't actually need this info, but I want one row per Id
  left_join(Fold_changes) %>% 
  mutate(ChrType=ifelse(Chr=='chrX', 'chrX', ifelse(Chr=='chrY', 'chrY', 'autosome')))

PlotFC_by_ChrType(all_DE_by_week, 'All DE')
all_DE_by_week %>% filter(ChrType != 'chrY' & SYMBOL != 'XIST') %>%
  PlotFC_by_ChrType("Non Y/XIST DE")


```

```{r}
p1 <- all_DE %>%
    mutate(ChrType=ifelse(Chr=='chrX', 'chrX', ifelse(Chr=='chrY', 'chrY', 'autosome')), Sex = ifelse(log2FoldChange < 0, 'F', 'M')) %>%
  ggplot(aes(x=Sex, fill=ChrType)) + geom_bar(position='stack') + 
    tufte_theme() +
    scale_y_continuous(breaks=seq(100,1000,100)) +
    scale_fill_manual(values=brewer.pal(6, "Set1")[c(1,2,4)]) +
    theme(legend.position='right') +
    theme(strip.switch.pad.grid = unit(10, 'cm')) +
    ylab('Number DE genes (FDR<0.1)') +
    xlab('Fetal Sex') +
    ggtitle("All Ages (12-19 PCW)")
  
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/AllSigBar.png", bg="transparent", width=128, height=96, units='mm', res=227)
p1
dev.off()
p1

p2 <- Fold_changes %>% filter(sig == 1) %>%
  mutate(week = paste(ifelse(week > 16, 
                      '15-16',
                      ifelse(week >= 15, 
                             '17-19', 
                             week)), 'PCW'),
         Sex = ifelse(log2FoldChange < 0, 'F', 'M')) %>% 
  left_join(dplyr::select(MalevsFemale_complete, Id, ChrType)) %>%
  ggplot(aes(x=Sex, fill=ChrType)) + geom_bar(position='stack') + 
  facet_grid(. ~ week) +
    tufte_theme() +
    scale_y_continuous(breaks=seq(100,1000,100)) +
    scale_fill_manual(values=brewer.pal(6, "Set1")[c(1,2,4)]) +
    theme(legend.position='right') +
    theme(strip.switch.pad.grid = unit(10, 'cm')) +
    ylab('Number DE genes (FDR<0.1)') +
    xlab('Fetal Sex')
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/ByWeekSigBar.png", bg="transparent", width=128, height=96, units='mm', res=227)
p2
dev.off()
p2

```

Fold differences for all genes (this one is slow)
```{r}
p1 <- MalevsFemale_complete %>% 
  dplyr::select(Id, ChrType) %>%
  inner_join(Fold_changes, by='Id') %>%
  PlotFC_by_ChrType("All genes CMP >= 1 in 45 or more samples")
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/AllGenesFC.png", bg="transparent", width=128, height=96, units='mm', res=227)
p1
dev.off()
p1
p2<-MalevsFemale_complete %>% 
  dplyr::select(Id, ChrType) %>%
  filter(ChrType != 'chrY' & Id != 'ENSG00000229807') %>% 
  inner_join(Fold_changes, by='Id') %>%
  PlotFC_by_ChrType("All Non Y/XIST genes")

png(file="~/BTSync/FetalRNAseq/LabNotes/Results/AllNonYGenesFC.png", bg="transparent", width=128, height=96, units='mm', res=227)
p2
dev.off()
p2
```

Figure 3: Expression data for selected genes characteristic of mature neurons (GABRA1, GRIN2A) and proliferating neural progenitor cells (NES, EOMES) in 14 PCW brain samples (8 male, 10 female).
```{r}
Week14 <- bind_rows(read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_edgeR/tables/MalevsFemale.down.txt",
                     "\t", escape_double = FALSE, trim_ws = TRUE),
                    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_edgeR/tables/MalevsFemale.up.txt",
                               "\t", escape_double = FALSE, trim_ws = TRUE) )%>% 
  dplyr::select(Id, starts_with('norm'), Female, Male) %>%
  separate(Id, c("Id"), sep='[.]', extra='drop')

makePlot <- function(letter, geneID) {
  Ensembl_Id <- bitr(geneID, fromType="SYMBOL", toType="ENSEMBL", OrgDb="org.Hs.eg.db")[,2] %>%
     paste(collapse = '_')
  PlotTimepoint(geneID, filter(Week14, Id == Ensembl_Id), letter)
}
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/Fig3.png", bg="transparent", width=210, height=140, units='mm', res=227)
grid.arrange(makePlot('A', 'GABRA1'), makePlot('B', 'GRIN2A'), makePlot('C', 'NES'), makePlot('D', 'EOMES'), ncol=2)
dev.off()
grid.arrange(makePlot('A', 'GABRA1'), makePlot('B', 'GRIN2A'), makePlot('C', 'NES'), makePlot('D', 'EOMES'), ncol=2)
```

Fold changes for Gene Set genes
```{r}
gene_sets <- GetGeneSets()
gene_sets <- dplyr::select(MalevsFemale_complete, ENSEMBL = Id, ChrType) %>%
  right_join(gene_sets)

Neuronal <- gene_sets %>% 
  filter(set=='Neuronal' & gene_name != 'XIST') %>%
  left_join(Fold_changes, by=c('ENSEMBL' = 'Id')) %>%
  mutate(Id = ENSEMBL)
NeuronalFC <- PlotFC(Neuronal, 'Neuronal Genes')
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/Neuronal.png", bg="transparent", width=128, height=96, units='mm', res=227)
NeuronalFC
dev.off()
NeuronalFC

NPC <- gene_sets %>% 
  filter(set=='NPC' & gene_name != 'RPS4Y1') %>%
  left_join(Fold_changes, by=c('ENSEMBL' = 'Id')) %>%
  mutate(Id = ENSEMBL)
NPCFC <- PlotFC(NPC, 'Neural Progenitor Cell Genes')
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/NPC.png", bg="transparent", width=128, height=96, units='mm', res=227)
NPCFC
dev.off()
NPCFC

RG <- gene_sets %>% 
  filter(set=='RadialGlia' & gene_name != 'RPS4Y1') %>%
  left_join(Fold_changes, by=c('ENSEMBL' = 'Id')) %>%
  mutate(Id = ENSEMBL)
RGFC<-PlotFC(RG, 'Radial Glia Genes')
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/RadialGlia.png", bg="transparent", width=128, height=96, units='mm', res=227)
RGFC
dev.off()
RGFC

AN <- gene_sets %>% 
  filter(set=='cluster7' & gene_name != 'RPS4Y1') %>%
  left_join(Fold_changes, by=c('ENSEMBL' = 'Id')) %>%
  mutate(Id = ENSEMBL)
ANFC<-PlotFC(AN, 'Top 21 Adult Neuron Genes (cluster 7)')
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/AdultNeuron.png", bg="transparent", width=128, height=96, units='mm', res=227)
ANFC
dev.off()
ANFC

FR <- gene_sets %>% 
  filter(set=='cluster9' & gene_name != 'RPS4Y1') %>%
  left_join(Fold_changes, by=c('ENSEMBL' = 'Id')) %>%
  mutate(Id = ENSEMBL)
FRFC<-PlotFC(FR, 'Top 21 Replicating Fetal Genes (cluster 9)')
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/FetalReplicating.png", bg="transparent", width=128, height=96, units='mm', res=227)
FRFC
dev.off()
FRFC
```

Figure 2: Log2 fold sex differences across time-points for (a) replicating fetal genes and b) adult neuron genes. Red dots indicate significant sex bias in expression (FDR < 0.1)
``` {r}
png(file="~/BTSync/FetalRNAseq/LabNotes/Results/Fig2.png", bg="transparent", width=210, 
    height=70, units='mm', res=227)
grid.arrange(PlotFC(AN, 'A.'), PlotFC(FR, 'B.'), ncol=2)
dev.off()

```

Week 14
```{r}

disease2gene <- gene_sets[, c("set", "gene_name")] 

FemaleDEG <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_edgeR/tables/FemaleUp.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

FemaleEnrichment <- summary(enricher(FemaleDEG$GeneID, TERM2GENE=disease2gene))
FemaleEnrichment$Direction <- 'FemaleBiased'

MaleDEG <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_edgeR/tables/MaleUp.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

MaleEnrichment <- summary(enricher(MaleDEG$GeneID, TERM2GENE=disease2gene))
MaleEnrichment$Direction <- 'MaleBiased'
Enrichment <- bind_rows(FemaleEnrichment, MaleEnrichment) %>% 
  separate(GeneRatio, c('x1', 'y1'), sep='/', remove=FALSE) %>% 
  separate(BgRatio, c('x2', 'y2'), sep='/', remove=FALSE) %>%
  mutate(Enrichment=(as.numeric(x1)/as.numeric(y1))/(as.numeric(x2)/as.numeric(y2))) %>%
  dplyr::select(Direction, ID, GeneRatio, BgRatio, Enrichment, pvalue, p.adjust, qvalue, geneID)
write_tsv(Enrichment, "~/BTSync/FetalRNAseq/LabNotes/Results/Enrichment14.txt")
Enrichment
```

All Weeks
```{r}
disease2gene <- gene_sets[, c("set", "gene_name")] 

FemaleDEG <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/FemaleUp12_19.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

FemaleEnrichment <- summary(enricher(FemaleDEG$GeneID, TERM2GENE=disease2gene))
FemaleEnrichment$Direction <- 'FemaleBiased'

MaleDEG <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/MaleUp12_19.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

MaleEnrichment <- summary(enricher(MaleDEG$GeneID, TERM2GENE=disease2gene))
#MaleEnrichment$Direction <- 'MaleBiased'
Enrichment <- bind_rows(FemaleEnrichment, MaleEnrichment) %>% 
  separate(GeneRatio, c('x1', 'y1'), sep='/', remove=FALSE) %>% 
  separate(BgRatio, c('x2', 'y2'), sep='/', remove=FALSE) %>%
  mutate(Enrichment=(as.numeric(x1)/as.numeric(y1))/(as.numeric(x2)/as.numeric(y2))) %>%
  dplyr::select(Direction, ID, GeneRatio, BgRatio, Enrichment, pvalue, p.adjust, qvalue, geneID)
write_tsv(Enrichment, "~/BTSync/FetalRNAseq/LabNotes/Results/EnrichmentAll.txt")
Enrichment
```

- Very low/negative fold-change correlation between weeks (when Y/XIST excluded)
- For some reason, week 17-19 is correlated with everything
- When week 14 is split into two samples, the second set is highly correlated with the full set (r=0.904), but the correlation with the first set is much weeker (r=0.675), and there are some serious outliers (one sample with log2FC=-10 for this analysis vs close to 0 for all others)


```{r}
Excluded <- MalevsFemale_complete %>% filter(ChrType == 'chrY' | Id == 'ENSG00000229807') %>%
  dplyr::select(Id)
W14_DE <-
FC_ByWeek <- Fold_changes %>%
  dplyr::select(-sig) %>%
  spread(week, log2FoldChange) %>% 
  dplyr::select(Id, 'w12'=`12`, 'w13'=`13`, 'w14'=`14`, 'w15_16'=`15.5`, 'w17_19'=`18`) %>%
  full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_19043_17921_19031_17221_18015_17053_17071_18529_18856_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::select(Id, w14_1= log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
  full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_17812_18282_18349_18355_18687_16428_16488_16640_16929_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  dplyr::select(Id, w14_2=log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
  full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_17812_18282_18349_17221_18015_16428_16488_18529_18856_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  dplyr::select(Id, w14_3=log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
  full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_19043_17921_19031_18355_18687_17053_17071_16640_16929_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  dplyr::select(Id, w14_4=log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
  anti_join(Excluded) 
dplyr::select(FC_ByWeek, -Id, -starts_with('w14_')) %>%
  ggpairs()
dplyr::select(FC_ByWeek, starts_with('w14')) %>%
  ggpairs()
filter(Fold_changes, week==14 & sig==1) %>% 
  inner_join(FC_ByWeek) %>% 
  dplyr::select(starts_with('w14')) %>% 
  ggpairs()
```

- For some reason, the second set is getting some genes with relatively small differences between Male and Female counts, but very large fold changes
- The fold changes are more complex estimates, so simply dividing Male by Female doesn't give the exact same numeber (and in fact I had to add 1 to deal with division by zero errors)
- However, in the case of the second set, fold changes calculated this way correlate highely with the original estimates (r=0.903 with no major outliers)
- The correlation is much lower in the case of the first set (r=.591, with a point at 0,-10)
- In the case of the second set, the original FC values correlate better with the full set than the recalculated ones (r=0.904 vs. r=0.847), while in the case of set 1, the correlation is lower (r=0.675 vs. r=0.733)
- It seems pretty clear that something went horribly wrong in the first run 

```{r}
recalculated <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_19043_17921_19031_17221_18015_17053_17071_18529_18856_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  separate(Id, c('Id'), sep = "[.]", extra='drop') %>%
  mutate(w14_1_2 = log2((Male+1)/(Female+1))) %>%
  dplyr::select(Id, w14_1_2) %>%
  inner_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_17812_18282_18349_18355_18687_16428_16488_16640_16929_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  separate(Id, c('Id'), sep = "[.]", extra='drop') %>%
  mutate(w14_2_2 = log2((Male+1)/(Female+1))) %>%
  dplyr::select(Id, w14_2_2)
  ) %>%
  inner_join(FC_ByWeek)

recalculated %>% dplyr::select(w14, w14_1, w14_1_2, w14_2, w14_2_2) %>%
    ggpairs()

```

- The correlation with the replication set is ...not good
- It's similar to the between-weeks values

```{r}
HDBRexpression <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBRexpression_FDR_0.1_edgeR/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% dplyr::select(Id, w14_rep=log2FoldChange) %>% separate(Id, c('Id'), sep = "[.]", extra='drop') %>%
  inner_join(recalculated) 

HDBRexpression %>% dplyr::select(w14, w14_1_2, w14_2, w14_rep) %>%
    ggpairs()

```


```{r}
DEseq_FC <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_DEseq/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::select(Id, w14= log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop') %>%
  full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_19043_17921_19031_17221_18015_17053_17071_18529_18856_FDR_0.1_DEseq/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::select(Id, w14_1= log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
  full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_18282_18349_18355_18687_16428_16488_16640_16929_FDR_0.1_DEseq/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  dplyr::select(Id, w14_2=log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
    full_join(
    read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_excl_17921_FDR_0.1_DEseq/tables/MalevsFemale.complete.txt",
"\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  dplyr::select(Id, w14_3=log2FoldChange) %>%
  separate(Id, c('Id'), sep = "[.]", extra='drop')
  ) %>%
  anti_join(Excluded) 
dplyr::select(DEseq_FC, -Id) %>%
  ggpairs()
```

``` {r}

IDgenes <- data.frame(SYMBOL=c('NLGN4X', 'RPS6KA3', 'ARX', 'PQBP1', 'KDM5C', 'OPHN1', 'NLGN3', 'SLC16A2', 'ATRX', 'AGTR2', 'MECP2', 'SLC6A8', 'IL1RAPL1', 'ZNF41', 'FTSJ1', 'TSPAN7', 'DLG3', 'PAK3', 'ACSL4', 'ARHGEF6', 'FMR1', 'AFF2', 'GDI1'))


XLID <- read_csv("~/BTSync/FetalRNAseq/LabNotes/Datafiles/XLID.txt", 
     col_names = FALSE) %>% dplyr::rename(SYMBOL=X1)

XLID_all <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/BG12_19.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% right_join(XLID) 
XLID_all %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/Results/week12_19_ID.txt")
XLID_all_sigF <- filter(XLID_all, padj < 0.1 & log2FoldChange < 0)
XLID_all_sigF$SYMBOL

XLID_all_sigM <- filter(XLID_all, padj < 0.1 & log2FoldChange > 0)

XLID_all_sigM$SYMBOL


XLID_14 <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/BG14.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% right_join(XLID)  

XLID_14 %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/Results/week14_ID.txt")
XLID_14_sigF <- filter(XLID_14, padj < 0.1 & log2FoldChange < 0)
XLID_14_sigF$SYMBOL
XLID_14_sigM <- filter(XLID_14, padj < 0.1 & log2FoldChange > 0)
XLID_14_sigM$SYMBOL

```

``` {r}
XLASD <- read_csv("~/BTSync/FetalRNAseq/LabNotes/Datafiles/XLASD.txt", 
     col_names = FALSE) %>% dplyr::rename(SYMBOL=X1)

XLASD_all <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/BG12_19.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% right_join(XLASD) 
XLASD_all %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/Results/week12_19_ASD.txt")
XLASD_all_sigF <- filter(XLASD_all, padj < 0.1 & log2FoldChange < 0)
XLASD_all_sigF$SYMBOL

XLASD_all_sigM <- filter(XLASD_all, padj < 0.1 & log2FoldChange > 0)

XLASD_all_sigM$SYMBOL


XLASD_14 <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/BG14.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% right_join(XLASD)  

XLASD_14 %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/Results/week14_ASD.txt")
XLASD_14_sigF <- filter(XLASD_14, padj < 0.1 & log2FoldChange < 0)
XLASD_14_sigF$SYMBOL
XLASD_14_sigM <- filter(XLASD_14, padj < 0.1 & log2FoldChange > 0)
XLASD_14_sigM$SYMBOL
```

``` {r}
Xinactivation <- read_tsv("~/BTSync/FetalRNAseq/LabNotes/Datafiles/Zheng13.txt")

FemaleUpX <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/FemaleUp12_19.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% filter(Chr == 'chrX') %>% dplyr::select(SYMBOL=GeneID, FemaleUp=Chr)
MaleUpX <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/MaleUp12_19.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% filter(Chr == 'chrX') %>% dplyr::select(SYMBOL=GeneID, MaleUp=Chr)
BG <- read_delim("~/BTSync/FetalRNAseq/LabNotes/Results/BG12_19.txt", 
     "\t", escape_double = FALSE, trim_ws = TRUE) %>% dplyr::select(SYMBOL, Id)

OtherStudy <- data.frame(SYMBOL =c("AKAP17A", "CD99", "stSG15779", "EIF1AX", "stSG1369", "stSG9723", "ARSD", "STS", "PNPLA4", "TBL1X", "OFD1", "TRAPPC2", "PIR", "INE2", "stSG4551", "GRPR", "RBBP7", "EIF2S3", "MED14", "USP9X", "ZNFXP", "KDM6A", "UBA1", "CDK16", "INE1", "KDM5C", "ADS56", "DXS1000E", "SMC1A", "ADS13", "AA383258", "RPS4X", "WI-12682"), Carrel=1) 


Results <- full_join(FemaleUpX, MaleUpX) %>% 
  full_join(Xinactivation, by=c('SYMBOL' = 'Genes')) %>%
  full_join(OtherStudy) %>% 
  left_join(BG) %>%
  mutate(Expressed = ifelse(! is.na(FemaleUp), 
                            'FemaleUp', 
                            ifelse(! is.na(MaleUp),
                                   'MaleUp',
                                   ifelse(is.na(Id), 
                                           'NotExpressed', 
                                           'Equal'))),
         Carrel = ifelse(is.na(Carrel), 0, 1),
         Reported = ifelse(is.na(Reported), 'NA', Reported)) %>%
  dplyr::select(-FemaleUp, -MaleUp)
write_delim(Results, "~/BTSync/FetalRNAseq/LabNotes/Results/Xinactivation.txt", delim='\t')

Results %>%  dplyr::select(Reported, Expressed, Carrel) %>% 
  table()

```

```{r}
my_db <- src_mysql("FetalRNAseq", host="localhost", user="root")
geneIDs <- collect(tbl(my_db, sql("SELECT DISTINCT GencodeFeatures.value AS 'GeneID' FROM 
GencodeFeatures, GencodeGTF WHERE 
GencodeGTF.id = GencodeFeatures.id AND 
GencodeFeatures.feature = 'gene_name' AND 
GencodeFeatures.id IN (SELECT GencodeGTF.id FROM GencodeGTF, GencodeFeatures WHERE 
                                             GencodeGTF.id = GencodeFeatures.id AND GencodeGTF.seqid = 'chrX')")))

SexBiasedGenes_14PCW <- read_csv("~/BTSync/FetalRNAseq/LabNotes/DMPs/SexBiasedGenes_14PCW.csv")

#bind_cols(duplicates, GetGeneIDs(paste(SexBiasedGenes_14PCW$Id, duplicates$version, sep='.')))

inner_join(SexBiasedGenes_14PCW, geneIDs, by=c('x' = 'GeneID')) %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/DMPs/SexBiasedX_14PCW.csv")
anti_join(SexBiasedGenes_14PCW, geneIDs, by=c('x' = 'GeneID')) %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/DMPs/SexBiasedAutosome_14PCW.csv")

SexBiasedGenes_All <- read_csv("~/BTSync/FetalRNAseq/LabNotes/DMPs/SexBiasedGenes_AllSamples.csv")
inner_join(SexBiasedGenes_All, geneIDs, by=c('x' = 'GeneID')) %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/DMPs/SexBiasedX_All.csv")
anti_join(SexBiasedGenes_All, geneIDs, by=c('x' = 'GeneID')) %>% write_tsv("~/BTSync/FetalRNAseq/LabNotes/DMPs/SexBiasedAutosome_All.csv")

```


```{r}
FemaleDEG <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_edgeR/tables/MalevsFemale.down.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)
FemaleDEG<-FemaleDEG %>% 
  dplyr::select(Id, starts_with('norm')) %>%
  gather(label, norm_count, -Id) %>%
  mutate(label=gsub('norm.', '', label )) %>%
  left_join(target)
MaleDEG <- read_delim("~/BTSync/FetalRNAseq/Counts/MvsF_14_HDBR_FDR_0.1_edgeR/tables/MalevsFemale.up.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)
MaleDEG<-MaleDEG %>% 
  dplyr::select(Id, starts_with('norm')) %>%
  gather(label, norm_count, -Id) %>%
  mutate(label=gsub('norm.', '', label )) %>%
  left_join(target)


cor <- bind_rows(FemaleDEG, MaleDEG) %>% group_by(Id) %>% summarise(r=cor.test(norm_count, RIN)[[4]], p_val=cor.test(norm_count, RIN)[[3]])
filter(cor, p_val < 0.05) %>% nrow()/nrow(cor)
ggplot(cor, aes(x=r)) + geom_histogram() + tufte_theme()
ggplot(cor, aes(x=p_val)) + geom_histogram() + tufte_theme()
```
```{r}
null <- data.frame(Id=rep(seq(1000), 105),
                   x=rnorm(105000),
                   y=rnorm(105000))
null_cor <- null %>% group_by(Id) %>% summarise(r=cor.test(x, y)[[4]], p_val=cor.test(x, y)[[3]])

ggplot(null_cor, aes(x=r)) + geom_histogram() + tufte_theme()
ggplot(null_cor, aes(x=p_val)) + geom_histogram() + tufte_theme()

```