---
title: "Figures"
author: "Heath O'Brien"
date: "4/9/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(fig.width=12, fig.height=8, fig.path='Figs/',
                      echo=FALSE, 
                      warning=FALSE, 
                      message=FALSE)

library(tidyverse)
library(stringr)
#library(gridExtra)
#library(clusterProfiler)
library(xlsx)
library(EnsDb.Hsapiens.v86)
library(ggbio)
library(AnnotationFilter)
source("../Shiny/GENEX-FB1/FormatGGplot.R")

gene_info <- read_delim("../Data/genes.txt", "\t", escape_double = FALSE, trim_ws = TRUE)

sample_info <- read_delim("../Data/SampleInfo.txt", "\t", col_types=cols(Sample='c'), escape_double = FALSE, trim_ws = TRUE) %>%
  mutate(label=as.character(Sample))


fittedBias <- read_delim("../Shiny/GENEX-FB1/Data/fitted.txt", "\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::rename(log2FoldDiff = log2FoldChange) %>%
  mutate(Sex = ifelse(log2FoldDiff < 0, 'F', 'M'),
         ChrType = ifelse(Chr %in% c('chrX','chrY'), Chr, 'autosome')) 
counts <- read_delim("../Shiny/GENEX-FB1/Data/counts12_20.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
counts_tr <- read_delim("../Shiny/GENEX-FB1/Data/counts12_20_tr.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
fittedBias_tr <- read_delim("../Shiny/GENEX-FB1/Data/fitted_tr.txt", "\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::rename(log2FoldDiff = log2FoldChange) %>%
  mutate(Sex = ifelse(log2FoldDiff < 0, 'F', 'M'),
         ChrType = ifelse(Chr %in% c('chrX','chrY'), Chr, 'autosome'))

fittedBias_topIsoform <- fittedBias_tr %>% group_by(GeneId) %>% arrange(padj) %>% dplyr::slice(1) %>% ungroup()

DEgenes <- filter(fittedBias, padj<=.1) %>% dplyr::select(Id)
DEtranscripts <- filter(fittedBias_topIsoform, padj<=.1 & !is.na(GeneId)) %>% dplyr::select(Id=GeneId) 
DEall <- inner_join(DEgenes, DEtranscripts) %>% 
  mutate(level='both') %>%
  full_join(DEgenes) %>%
  mutate(level=ifelse(is.na(level), 'gene-level', level)) %>%
  full_join(DEtranscripts) %>%
  mutate(level=ifelse(is.na(level), 'transcript-level', level)) %>%
  inner_join(gene_info, by=c('Id' = 'gene_id')) %>%
  left_join(dplyr::select(fittedBias, Id, gene_padj=padj)) %>%
  left_join(dplyr::select(fittedBias_topIsoform, Id=GeneId, transcript_padj=padj, Sex))
```


Figure 1.

```{r}
PARgenes <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/PARgenes.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

DEall <- left_join(DEall, dplyr::select(PARgenes, gene_name=`Approved Symbol`, PAR=Unknown)) %>%
  mutate(ChrType=ifelse(is.na(PAR), ChrType, 'PAR'))
Fig1 <- ggplot(DEall, aes(x=Sex, fill=factor(ChrType, levels=c('autosomal', 'PAR', 'chrX', 'chrY')))) + geom_bar(position='stack') + 
  facet_grid(~ level) +
    figure_theme() +
    #scale_y_continuous(breaks=seq(100,1000,100)) +
    scale_fill_manual(values=brewer.pal(6, "Set1")[c(1,6,4,2)]) +
    theme(strip.switch.pad.grid = unit(10, 'cm')) +
    ylab('Number sex-biased genes (FDR<0.1)') +
    xlab('Fetal sex') +
    theme(axis.title.y=element_text(size=9)) +
    theme(axis.title.x=element_text(size=9)) +
    ggtitle('A.') +
    theme(legend.position = 'right')
png(file="../Figures/Fig1.png", bg="transparent", width=183, 
    height=70, units='mm', res=300)
Fig1
dev.off()
pdf(file="../Figures/Fig1.pdf", 7.2, 3)
Fig1
dev.off()
Fig1
```

Figure 2: Log2 fold sex differences across time-points for (a) replicating fetal genes and b) adult neuron genes. Red dots indicate significant sex bias in expression (FDR < 0.1)
``` {r}
txdb <- EnsDb.Hsapiens.v86
KDM2B_tr1 <- exonsBy(txdb, filter=TxIdFilter('ENST00000538046')) 
p1 <- autoplot(KDM2B_tr1, group.selfish = TRUE, names='')
KDM2B_tr2 <- exonsBy(txdb, filter=TxIdFilter('ENST00000611216')) 
p2 <- autoplot(KDM2B_tr2, group.selfish = TRUE)
p3 <- autoplot(txdb, which = GeneIdFilter('ENSG00000089094'), stat="reduce")
gene_models <- tracks(ENST00000538046 = p1, ENST00000611216=p2, gene = p3)
#pdf("Fig2A.pdf")
gene_models+main_theme()
ggsave("../Figures/Fig2A.pdf")
#dev.off()


KDM2Bmean <- filter(fittedBias_tr, Id %in% c('ENST00000611216', 'ENST00000538046')) %>%
  bind_rows(filter(fittedBias, Id == 'ENSG00000089094')) %>%
  dplyr::select(Male, Female, Id, qval=padj) %>%
  gather('Sex', 'mean', -Id, -qval) %>%
  mutate(facet=factor(paste0(Id, '\nFDR=', signif(qval, digits = 3))))

KDM2Bmean$facet<-factor(KDM2Bmean$facet, levels=c(levels(KDM2Bmean$facet)[2], levels(KDM2Bmean$facet)[3], levels(KDM2Bmean$facet)[1]))

  
KDM2Bcounts <- filter(counts_tr, Id %in% c('ENST00000611216', 'ENST00000538046')) %>%
  bind_rows(filter(counts, Id=='ENSG00000089094')) %>%
  dplyr::select(-SYMBOL, -GeneId, -Chr, -ChrType) %>%
  gather(key, value, -Id, -gene_type) %>%
  separate(key, into=c('norm', 'Sample'), sep='[.]') %>%
  dplyr::select(Sample, value, Id) %>%
  left_join(sample_info) %>%
  left_join(dplyr::select(KDM2Bmean, Id, qval) %>% group_by(Id) %>% dplyr::slice(1)) %>%
  mutate(facet=factor(paste0(Id, '\nFDR=', signif(qval, digits = 3)))) %>%
  filter(Sample != '16115' | Id != 'ENST00000538046') # filter count outlier

KDM2Bcounts$facet<-factor(KDM2Bcounts$facet, levels=c(levels(KDM2Bcounts$facet)[2], levels(KDM2Bcounts$facet)[3], levels(KDM2Bcounts$facet)[1]))

ggplot(KDM2Bcounts, aes(x=Sex, colour=Sex)) + 
    geom_errorbar(aes(ymin=mean, ymax=mean), colour='black', size=.5, width=.5, data=KDM2Bmean) +
    geom_jitter(aes(y=value), height = 0, width=.1, alpha=.75, size=1) + 
    facet_grid(facet ~ ., scales="free") +
    ylab("Normalized Counts") +
    xlab('') +
    main_theme() +
    scale_y_continuous(position = "right") +
    scale_colour_brewer(type = "qual", palette = 6) +
    theme(
      strip.background = element_blank(),
      strip.text = element_blank()
    )
ggsave("../Figures/Fig2B.pdf")
#pdf("../Figures/Fig2B.pdf")
#ANOS1expression
#dev.off()

```

Figure S1

```{r}
p1 <- ggplot(target, aes(x=PCW, fill=Sex)) +
  geom_bar() +
  facet_grid(Sex ~ .) +
  scale_y_continuous(breaks=seq(0,100, 2)) +
  scale_x_continuous(breaks=seq(11,19, 1)) +
  figure_theme() +
  xlab("Post-conception weeks") +
  ylab("Count") +
  theme(axis.text.x=element_text(size=8)) +
  scale_fill_brewer(type = "qual", palette = 6) +
  ggtitle('A.')

target <- mutate(target, age_bin =ifelse(target$PCW == 16, 15, 
                                     ifelse(target$PCW > 16, 16, target$PCW)))
p2 <- ggplot(target, aes(x=age_bin, fill=Sex)) +
  geom_bar() +
  facet_grid(Sex ~ .) +
  scale_y_continuous(breaks=seq(0,100, 2)) +
  scale_x_continuous(breaks=c(11,12,13,14,15,16),
                     labels=c('11','12','13','14','15-16','17-19')) +
  figure_theme() +
  xlab("Post-conception weeks") +
  ylab("Count") +
  theme(axis.text.x=element_text(size=8)) +
  scale_fill_brewer(type = "qual", palette = 6) +
  ggtitle('B.')

pdf(file="Figures/FigS1.pdf", 7.2, 5.5)
grid.arrange(p1, p2, ncol=2)
dev.off()
png(file="Figures/FigS1.png", bg="transparent", width=183, height=140, units='mm', res=300)
grid.arrange(p1, p2, ncol=2)
dev.off()
grid.arrange(p1, p2, ncol=2)

```

Figure S2
```{r}
Neuronal <- gene_sets %>% 
  filter(set=='Neuronal' & gene_name != 'XIST') %>% 
  inner_join(fittedBias)
group_by(Neuronal, ageBin) %>% summarise(n=n())

RG <- gene_sets %>% 
  filter(set=='RadialGlia' & gene_name != 'XIST') %>%
  inner_join(fittedBias)
group_by(RG, ageBin) %>% summarise(n=n())

NPC <- gene_sets %>% 
  filter(set=='NPC' & gene_name != 'RPS4Y1') %>%
  inner_join(fittedBias)
group_by(NPC, ageBin) %>% summarise(n=n())

png(file="Figures/FigS2.png", bg="transparent", width=183, 
    height=210, units='mm', res=300)
grid.arrange(PlotFC(NPC, 'A. iPSC-derived neural progenitor cell genes'),
             PlotFC(RG, 'B. Radial glia genes'),
             PlotFC(Neuronal, 'C. Fetal neuron genes'), 
             ncol=1)
dev.off()
pdf(file="Figures/FigS2.pdf", 7.2, 9)
grid.arrange(PlotFC(NPC, 'A. iPSC-derived neural progenitor cell genes'),
             PlotFC(RG, 'B. Radial glia genes'),
             PlotFC(Neuronal, 'C. Fetal neuron genes'), 
             ncol=1)
dev.off()

grid.arrange(PlotFC(NPC, 'A. iPSC-derived neural progenitor cell genes'),
             PlotFC(RG, 'B. Radial glia genes'),
             PlotFC(Neuronal, 'C. Fetal neuron genes'), 
             ncol=1)

```

Enrichment Analysis
Week 14
```{r}

#disease2gene <- filter(gene_sets, ! set %in% c('SandersASD', 'SFARI'))[, c("set", "gene_name")] 
disease2gene <- filter(gene_sets, ! set %in% c('SandersASD', 'SFARI')) %>% dplyr::select(set, gene_name) 
FemaleDEG <- fittedBias %>% filter(ageBin == '14' & padj<0.1 & log2FoldDiff <0)

FemaleEnrichment <- summary(enricher(FemaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
FemaleEnrichment$Direction <- 'FemaleBiased'

MaleDEG <- fittedBias %>% filter(ageBin == '14' & padj<0.1 & log2FoldDiff >0)

MaleEnrichment <- summary(enricher(MaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
MaleEnrichment$Direction <- 'MaleBiased'
Enrichment <- bind_rows(FemaleEnrichment, MaleEnrichment) %>% 
  separate(GeneRatio, c('x1', 'y1'), sep='/', remove=FALSE) %>% 
  separate(BgRatio, c('x2', 'y2'), sep='/', remove=FALSE) %>%
  mutate(Enrichment=(as.numeric(x1)/as.numeric(y1))/(as.numeric(x2)/as.numeric(y2))) %>%
  dplyr::select(Direction, ID, GeneRatio, BgRatio, Enrichment, pvalue, p.adjust, qvalue, geneID)

#Need to correct for 10 test because we tested 2 sets of DE genes
#For some mysterious reason, enricher is correcting for 4 tests even though there are clearly 5 gene sets being tested.
Enrichment$p.adjust <- p.adjust(Enrichment$pvalue, method='bonferroni', n=10)

write_tsv(Enrichment, "Tables/Enrichment14.txt")
Enrichment
```

## All weeks (65 high-cofindence Autism genes and 79 high confidence X-linked ID genes)
- Sanders et al 2015, Piton et al 2013
```{r}

disease2gene <- bind_rowsfilter(gene_sets, set %in% c('SandersASD', 'PitonID'))[, c("set", "gene_name")]

FemaleDEG <- fittedBiasAll %>% filter(padj<0.1 & log2FoldDiff < 0)

FemaleEnrichment <- summary(enricher(FemaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
dim(FemaleEnrichment)
#FemaleEnrichment$Direction <- 'FemaleBiased'

MaleDEG <- fittedBiasAll %>% filter(padj<0.1 & log2FoldDiff > 0)

MaleEnrichment <- summary(enricher(MaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
MaleEnrichment$Direction <- 'MaleBiased'
Enrichment <- bind_rows(FemaleEnrichment, MaleEnrichment) %>% 
  separate(GeneRatio, c('x1', 'y1'), sep='/', remove=FALSE) %>% 
  separate(BgRatio, c('x2', 'y2'), sep='/', remove=FALSE) %>%
  mutate(Enrichment=(as.numeric(x1)/as.numeric(y1))/(as.numeric(x2)/as.numeric(y2))) %>%
  dplyr::select(Direction, ID, GeneRatio, BgRatio, Enrichment, pvalue, p.adjust, qvalue, geneID)

#Need to correct for test because 2 sets of DE genes
Enrichment$p.adjust <- p.adjust(Enrichment$pvalue, method='bonferroni', n=2)
write_tsv(Enrichment, "Tables/EnrichmentAll.txt")
Enrichment
```

Supplementary Table 1: Sampling info
```{r}
TotalFragments <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Tables/CountSummary.txt",
"\t", escape_double = FALSE, col_types = cols(Sample = col_character()),
trim_ws = TRUE) %>% dplyr::select(Sample, TotalFragments)

SampleInfo <- read_tsv("../Data/SampleInfo.txt", col_types = cols(Sample = col_character())) %>%
  full_join(TotalFragments) %>%
  mutate(`Total Read Pairs (millions)` = round(TotalFragments/1000000)) %>%
  dplyr::select(-TotalFragments)

write.xlsx2(as.data.frame(SampleInfo), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS1',
            append=FALSE)

```

Supplementary Table 2: Genes with significant sex Biases (FDR<0.1)
```{r}
DEG12_19 <- fittedBias %>% filter(padj<0.1 & !is.na(Chr)) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  mutate(Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl Gene ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG12_19), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS2',
            append=TRUE)
```


Supplementary Table 3: Transcripts with significant sex Biases (FDR<0.1)
```{r}
DEG12_19_tr <- fittedBias_tr %>% filter(padj<0.1 & !is.na(Chr)) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  mutate(Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl Transcript ID` = Id,
         `Ensembl Gene ID` = GeneId,
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG12_19_tr), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS3',
            append=TRUE)
```

# Supplementary Table 4: Comparison with previous findings
- transcript-level analysis = isoform with the lowset p-value

```{r}
previous <- read_tsv("../Data/previous_findings.txt")
sig_res <- fittedBias %>%
  filter(padj < 0.1 & !is.na(Chr)) %>%
  mutate(result = paste0(format(log2FoldDiff, digits=2), ' (', format(padj, digits=2), ')'),
         level = 'Gene-level Analysis') %>%
  dplyr::select(Id, SYMBOL, Chr, result, level)

sig_res_tr <- fittedBias_tr %>%
  filter(padj < 0.1 & !is.na(Chr)) %>%
  arrange(padj) %>%
  group_by(GeneId) %>%
  dplyr::slice(1) %>%
  dplyr::select(-Id, Id=GeneId) %>%
  mutate(result = paste0(format(log2FoldDiff, digits=2), ' (', format(padj, digits=2), ')'),
         level = 'Transcript-level Analysis') %>%
  dplyr::select(Id, SYMBOL, Chr, result, level)

sig_res_all <- bind_rows(sig_res, sig_res_tr) %>%
  spread(level, result) %>% 
  mutate(Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) %>%
  arrange(Chr) %>% 
  dplyr::rename(`Ensembl ID` = Id, 
                `Gene ID` = SYMBOL,
                Chromosome = Chr) %>%
  left_join(previous)

write.xlsx2(as.data.frame(sig_res_all), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS4',
            append=TRUE)
```

Supplementary Table 3
```{r}
Xinactivation <- read_tsv('Data/Xinactivation.txt')
write.xlsx2(as.data.frame(Xinactivation), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS3',
            append=TRUE)
```

Supplementary Tables 4 & 5
```{r}
GO_M_all <- read_tsv('Data/DAVID_M12_19.txt')
write.xlsx2(as.data.frame(GO_M_all), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS4',
            append=TRUE)

GO_F_all <- read_tsv('Data/DAVID_F12_19.txt')
write.xlsx2(as.data.frame(GO_F_all), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS5',
            append=TRUE)

```

Supplementary Tables 6 - 10
```{r}
DEG12 <- fittedBias %>% filter(padj<0.1 & ageBin == 12) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG12), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS6',
            append=TRUE)

DEG13 <- fittedBias %>% filter(padj<0.1 & ageBin == 13) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj) 

write.xlsx2(as.data.frame(DEG13), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS7',
            append=TRUE)

DEG14 <- fittedBias %>% filter(padj<0.1 & ageBin == 14) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG14), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS8',
            append=TRUE)

DEG15_16 <- fittedBias %>% filter(padj<0.1 & ageBin == '15-16') %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG15_16), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS9',
            append=TRUE)

DEG17_19 <- fittedBias %>% filter(padj<0.1 & ageBin == '17-19') %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG17_19), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS10',
            append=TRUE)
```

Supplementary Tables 11 & 12
```{r}
GO_M_14 <- read_tsv('Data/DAVID_M14.txt')
write.xlsx2(as.data.frame(GO_M_14), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS11',
            append=TRUE)

GO_F_14 <- read_tsv('Data/DAVID_F14.txt')
write.xlsx2(as.data.frame(GO_F_14), 
            file="Tables/SupplemtalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS12',
            append=TRUE)

```

