---
title: "Figures"
author: "Heath O'Brien"
date: "4/9/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_knit$set(fig.width=12, fig.height=8, fig.path='Figs/',
                      echo=FALSE, 
                      warning=FALSE, 
                      message=FALSE)

library(tidyverse)
library(stringr)
library(gridExtra)
library(clusterProfiler)
library(xlsx)
library(EnsDb.Hsapiens.v86)
library(ggbio)
library(AnnotationFilter)
source("../Shiny/GENEX-FB1/FormatGGplot.R")

gene_info <- read_delim("../Data/genes.txt", "\t", escape_double = FALSE, trim_ws = TRUE)

sample_info <- read_delim("../Data/SampleInfo.txt", "\t", col_types=cols(Sample='c'), escape_double = FALSE, trim_ws = TRUE) %>%
  mutate(label=as.character(Sample))


fittedBias <- read_delim("../Shiny/GENEX-FB1/Data/fitted.txt", "\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::rename(log2FoldDiff = log2FoldChange) %>%
  mutate(Sex = ifelse(log2FoldDiff < 0, 'F', 'M')) %>%
  filter(!is.na(padj) & !is.na(Chr))
counts <- read_delim("../Shiny/GENEX-FB1/Data/counts12_20.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
counts_tr <- read_delim("../Shiny/GENEX-FB1/Data/counts12_20_tr.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
fittedBias_tr <- read_delim("../Shiny/GENEX-FB1/Data/fitted_tr.txt", "\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::rename(log2FoldDiff = log2FoldChange) %>%
  mutate(Sex = ifelse(log2FoldDiff < 0, 'F', 'M')) %>%
  filter(!is.na(padj) & !is.na(Chr))

fittedPCW <- read_delim("../Shiny/GENEX-FB1/Data/dropPCW.txt", "\t", escape_double = FALSE, trim_ws = TRUE) %>%
  dplyr::rename(log2FoldDiff = log2FoldChange) %>%
  filter(!is.na(padj) & !is.na(Chr))


fittedBias_topIsoform <- fittedBias_tr %>% group_by(GeneId) %>% arrange(padj) %>% dplyr::slice(1) %>% ungroup()

DEgenes <- filter(fittedBias, padj<=.1) %>% dplyr::select(Id)
DEtranscripts <- filter(fittedBias_topIsoform, padj<=.1 & !is.na(GeneId)) %>% dplyr::select(Id=GeneId) 
DEall <- inner_join(DEgenes, DEtranscripts) %>% 
  mutate(level='both') %>%
  full_join(DEgenes) %>%
  mutate(level=ifelse(is.na(level), 'gene-level', level)) %>%
  full_join(DEtranscripts) %>%
  mutate(level=ifelse(is.na(level), 'transcript-level', level)) %>%
  inner_join(gene_info, by=c('Id' = 'gene_id')) %>%
  left_join(dplyr::select(fittedBias, Id, gene_padj=padj)) %>%
  left_join(dplyr::select(fittedBias_topIsoform, Id=GeneId, transcript_padj=padj, Sex))

XCIsummary <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/XCIsummary.txt", "\t", escape_double = FALSE, col_names = TRUE, trim_ws = TRUE)

GTExXCI <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/GTExXCI.txt", "\t", escape_double = FALSE, col_names = TRUE, trim_ws = TRUE)

GTExNovel <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/GTExNovel.txt", "\t", escape_double = FALSE, col_names = TRUE, trim_ws = TRUE)

NewFindings <- filter(GTExXCI, NewCall=='x') %>% 
  dplyr::rename(GeneName = Name) %>%
  mutate(GTExStatus = 'variable') %>%
  dplyr::select(GeneID, GeneName, Chr, Start, End, GTExStatus) %>%
  full_join(GTExNovel) %>%
  mutate(GTExStatus = ifelse(is.na(GTExStatus), 'escape', GTExStatus))

BallatonXCI <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/BallatonXCI.txt",
"\t", escape_double = FALSE, col_names = c("SYMBOL", "Status"),
trim_ws = TRUE)

BallatonXCI <- group_by(BallatonXCI, SYMBOL) %>% dplyr::summarise(n=n()) %>% filter(n==1) %>% left_join(BallatonXCI) %>% dplyr::select(-n)

XCI <- dplyr::select(NewFindings, GeneID, GeneName, Chr, Start, End, GTExStatus) %>%
  full_join(XCIsummary, by=c("GeneID", "GeneName", "Chr", "Start", "End")) %>%
  mutate(GTExStatus=ifelse(is.na(GTExStatus), CombinedXCIstatus, GTExStatus), Id=str_extract(GeneID, '^[^.]+')) %>%
  dplyr::select(-Chr)

XCI <- full_join(BallatonXCI, XCI, by=c('SYMBOL'='GeneName')) 
```

Figure 1.

```{r}
PlotExpression<-function(geneID, fig_num, counts, fittedPCW, sample_info) {
  fit_params <- filter(fittedPCW, SYMBOL==geneID)
  data <- counts %>% filter(SYMBOL == geneID) %>%  
    dplyr::select(-one_of('SYMBOL', 'Id', 'Chr', 'ChrType', 'gene_type')) %>%
    gather() %>%
    separate(key, into=c('norm', 'Sample'), sep='[.]') %>%
    dplyr::select(Sample, value) %>%
    left_join(sample_info)
  mean_age<-mean(sample_info$PCW)
  fit <- data.frame(PCW=seq(12,19)) %>%
    mutate(fit=fit_params$baseMean*2^(fit_params$log2FoldDiff*(PCW-mean_age)))
    title<-paste0(fig_num, ': ', geneID)
  plot<-  ggplot(data, aes(x=PCW, y=value)) + 
    geom_jitter(height = 0, width=.1, alpha=.75) + 
    geom_line(aes(y=fit), colour='black', data=fit) +
    scale_x_continuous(breaks=seq(12, 20)) +
    ylab("Normalized Counts") +
    xlab('Post-Conception Weeks') +
    main_theme() +
    scale_colour_brewer(type = "qual", palette = 6) +
    ggtitle(title) +
    theme(axis.title.x=element_text(size=12),
          axis.title.y=element_text(size=12))
  plot
}
pdf("../Paper/Fig1.pdf")
grid.arrange(PlotExpression('TTLL7', 'A', counts, fittedPCW, sample_info),
             PlotExpression('RHBDL3', 'B', counts, fittedPCW, sample_info), 
             ncol=2)
dev.off()
#ggsave("../Paper/Fig1.pdf")
```

Figure 2.

```{r}
PARgenes <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/PARgenes.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

DEall <- left_join(DEall, dplyr::select(PARgenes, gene_name=`Approved Symbol`, PAR=Unknown)) %>%
  mutate(ChrType=ifelse(is.na(PAR), ChrType, 'PAR'))

DEall %>% mutate(level=relevel(factor(ifelse(level=='both', 
                                      'Both gene and transcript level',
                                      ifelse(level=='gene-level', 
                                             'Additional at gene level only',
                                             ifelse(level=='transcript-level',
                                                    'Additional at transcript level only',
                                                    NA)))),
                                      ref='Both gene and transcript level')) %>%
  ggplot(aes(x=Sex, fill=factor(ChrType, levels=c('autosomal', 'PAR', 'chrX', 'chrY')))) + geom_bar(position='stack') + 
  facet_grid(~ level) +
    figure_theme() +
    #scale_y_continuous(breaks=seq(100,1000,100)) +
    scale_fill_manual(values=brewer.pal(6, "Set1")[c(1,6,4,2)]) +
    theme(strip.switch.pad.grid = unit(10, 'cm')) +
    ylab('Number sex-biased genes (FDR<0.1)') +
    xlab('Fetal sex') +
    theme(axis.title.y=element_text(size=9)) +
    theme(axis.title.x=element_text(size=9)) +
    theme(legend.position = 'right')
ggsave("../Paper/Fig2.pdf")
```

Figure 3: 
``` {r}
ensdb <- EnsDb.Hsapiens.v86
#KDM2B_tr1 <- exonsBy(ensdb, filter=TxIdFilter('ENST00000538046')) 
p1and2 <- autoplot(ensdb, TxIdFilter(c('ENST00000538046', 'ENST00000611216')), gap.geom = "chevron") +
  main_theme() +
  theme(axis.text.x=element_text(size=9))

p1 <- autoplot(ensdb, TxIdFilter('ENST00000538046'), gap.geom = "chevron", label=FALSE) +
  main_theme() +
  theme(axis.text.x=element_text(size=9))
p2 <- autoplot(ensdb, TxIdFilter('ENST00000611216'), gap.geom = "chevron", label=FALSE) +
  main_theme() +
  theme(axis.text.x=element_text(size=9))
#KDM2B_tr2 <- exonsBy(ensdb, filter=TxIdFilter('ENST00000611216')) 
p3 <- autoplot(ensdb, GeneIdFilter('ENSG00000089094'), stat="reduce") +
  main_theme() +
  theme(axis.text.x=element_text(size=9))
gene_models <- tracks(ENST00000538046 = p1, ENST00000611216=p2, gene = p3)
#gene_models <- tracks(Transcripts = p1and2, Gene = p3, heights=c(2,1))
#pdf("Fig2A.pdf")
gene_models
ggsave("../Figures/Fig3A.pdf")
#dev.off()


KDM2Bmean <- filter(fittedBias_tr, Id %in% c('ENST00000611216', 'ENST00000538046')) %>%
  bind_rows(filter(fittedBias, Id == 'ENSG00000089094')) %>%
  dplyr::select(Male, Female, Id, qval=padj) %>%
  gather('Sex', 'mean', -Id, -qval) %>%
  mutate(facet=factor(paste0(Id, '\nFDR=', signif(qval, digits = 3))))

KDM2Bmean$facet<-factor(KDM2Bmean$facet, levels=c(levels(KDM2Bmean$facet)[2], levels(KDM2Bmean$facet)[3], levels(KDM2Bmean$facet)[1]))

  
KDM2Bcounts <- filter(counts_tr, Id %in% c('ENST00000611216', 'ENST00000538046')) %>%
  bind_rows(filter(counts, Id=='ENSG00000089094')) %>%
  dplyr::select(-SYMBOL, -GeneId, -Chr, -ChrType) %>%
  gather(key, value, -Id, -gene_type) %>%
  separate(key, into=c('norm', 'Sample'), sep='[.]') %>%
  dplyr::select(Sample, value, Id) %>%
  left_join(sample_info) %>%
  left_join(dplyr::select(KDM2Bmean, Id, qval) %>% group_by(Id) %>% dplyr::slice(1)) %>%
  mutate(facet=factor(paste0(Id, '\nFDR=', signif(qval, digits = 3)))) %>%
  filter(Sample != '16115' | Id != 'ENST00000538046') # filter count outlier

KDM2Bcounts$facet<-factor(KDM2Bcounts$facet, levels=c(levels(KDM2Bcounts$facet)[2], levels(KDM2Bcounts$facet)[3], levels(KDM2Bcounts$facet)[1]))

ggplot(KDM2Bcounts, aes(x=Sex, colour=Sex)) + 
    geom_errorbar(aes(ymin=mean, ymax=mean), colour='black', size=.5, width=.5, data=KDM2Bmean) +
    geom_jitter(aes(y=value), height = 0, width=.1, alpha=.75, size=1) + 
    facet_grid(facet ~ ., scales="free") +
    ylab("Normalized Counts") +
    xlab('') +
    main_theme() +
    scale_y_continuous(position = "right") +
    scale_colour_brewer(type = "qual", palette = 6) +
    theme(
      strip.background = element_blank(),
      strip.text = element_blank()
    )
ggsave("../Figures/Fig3B.pdf")
#pdf("../Figures/Fig2B.pdf")
#ANOS1expression
#dev.off()

```

Figure 4:

```{r}

fittedBias <- fittedBias %>%  mutate(gene_type = ifelse(gene_type %in% c("antisense_RNA", "lincRNA", "protein_coding", "miRNA"),
                            gene_type, 
                            ifelse(gene_type %in% c("snRNA", "snoRNA"),
                                   "snRNA/snoRNA",
                                   ifelse(str_detect(gene_type, "pseudogene"), 
                                          "pseudogene", 
                                          "other"))))

fittedBias %>% filter(padj<.1 & ChrType == 'autosomal') %>%
  group_by(gene_type) %>% 
  summarise(n=n(), proportion = n/nrow(filter(fittedBias %>% filter(padj<.1 & ChrType == 'autosomal')))*100, DE='SexBiased') %>%
  bind_rows(fittedBias %>% filter(!is.na(padj) & ChrType == 'autosomal') %>%
              group_by(gene_type) %>% 
              summarise(n=n(), proportion = n/nrow(filter(fittedBias %>% filter(!is.na(padj) & ChrType == 'autosomal')))*100, DE='Background')) %>%
  mutate(gene_type = factor(str_replace(gene_type, '_', ' ')), 
         gene_type=factor(gene_type, levels=c(levels(gene_type)[levels(gene_type) != "other"], "other"))) %>%
  ggplot(aes(x=DE, fill=DE, y=proportion)) + 
  ggplot2::geom_bar(stat='identity') +
  facet_grid(. ~gene_type) +
  xlab('') +
  ylab('% gene type') +
  main_theme() +
  theme(legend.position = 'bottom',
        axis.text.x=element_blank()) +
  scale_fill_manual(values=brewer.pal(6, "Set1")[c(1,2)])

ggsave("../Paper/Fig4.pdf")

```

Figure 5:

```{r}
PlotExpression <- function(counts, sample_info, geneID) {
  data <- counts %>% filter(SYMBOL == geneID) %>%  
    dplyr::select(-one_of('SYMBOL', 'Id', 'Chr', 'ChrType', 'gene_type')) %>%
    gather() %>%
    separate(key, into=c('norm', 'Sample'), sep='[.]') %>%
    dplyr::select(Sample, value) %>%
    left_join(sample_info)
  plot<-  ggplot(data, aes(x=PCW, y=value, colour=Sex)) + 
    geom_jitter(height = 0, width=.1, alpha=.75) + 
    scale_x_continuous(breaks=seq(12, 20)) +
    ylab("Normalized Counts") +
    xlab('Post-Conception Weeks') +
    main_theme() +
    scale_colour_brewer(type = "qual", palette = 6) +
    theme(legend.position = 'bottom')
    ggtitle(geneID) 
  plot
}

PlotExpression(counts, sample_info, 'KDM5C') 

ggsave("../Paper/Fig5.pdf")
```

Figure S1
- PRKX is Male biased and not in the PAR. It is reported as Female biased in GTEx
- NLGN4X is discrepent (according to Ballaton, it was found to be 100% E by Carrel & Willard; according to Tukiainen it was not tested (VE according to Cotton et al.))
```{r}
XCI_DE <- left_join(XCI, fittedBias, by='SYMBOL') %>% 
  mutate(sig=ifelse(padj<.1, 1,0)) %>%
  filter(SYMBOL != 'XIST')
ggplot(XCI_DE, aes(x=Status, y=log2FoldDiff, colour=factor(sig))) +
    geom_jitter(height = 0, width=.25, alpha=.25) +
    facet_grid(GTExStatus ~ .) +
    main_theme() +
    scale_colour_manual(values=c("black", "red")) +
    ylab("Log2 Fold Difference") +
    theme(axis.text.x=element_text(size=9, angle=45, hjust=1),
          axis.text.y=element_text(size=9),
          axis.title.x=element_text(size=12),
          axis.title.y=element_text(size=12))
ggsave("../Paper/FigS1.pdf")   
```

Enrichment Analysis
```{r}
background <- fittedBias %>% mutate(set="Background", reference=NA) %>% 
  dplyr::select(gene_name=SYMBOL, set, reference, ENSEMBL=Id)

gene_sets <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/GeneSets.txt",
"\t", escape_double = FALSE, trim_ws = TRUE)

gene_sets <- fittedBias %>% filter(gene_type == 'lincRNA') %>% mutate(set="lincRNA", reference='biomart') %>% 
  dplyr::select(gene_name=SYMBOL, set, reference, ENSEMBL=Id) %>%
  bind_rows(gene_sets)

gene_sets <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/ERR1_Q2.txt",
"\t", escape_double = FALSE, col_names = colnames(gene_sets)[1:3],
trim_ws = TRUE) %>% left_join(dplyr::select(gene_info, gene_name, ENSEMBL=gene_id)) %>% bind_rows(gene_sets)

gene_sets_filtered <- semi_join(gene_sets, dplyr::select(background, ENSEMBL))

group_by(gene_sets, set) %>% 
  summarise(set_size=n()) %>%
  full_join(group_by(gene_sets_filtered, set) %>% summarise(expressed=n()))

gene_sets_filtered <- bind_rows(gene_sets_filtered, background)

as.data.frame(enricher(filter(DEall, (level=='both' | level=='gene-level') & ChrType=='autosomal')$Id, TERM2GENE=dplyr::select(gene_sets_filtered, set, ENSEMBL), pAdjustMethod='bonferroni', pvalueCutoff = .2, maxGSSize = 10000))

as.data.frame(enricher(filter(DEall, (level=='both') & ChrType=='autosomal')$Id, TERM2GENE=dplyr::select(gene_sets_filtered, set, ENSEMBL), pAdjustMethod='bonferroni', pvalueCutoff = .2, maxGSSize = 10000))

as.data.frame(enricher(filter(fittedBias, pvalue<.05 & ChrType=='autosomal' & log2FoldDiff>0)$Id, TERM2GENE=dplyr::select(gene_sets_filtered, set, ENSEMBL), pAdjustMethod='bonferroni', pvalueCutoff = .2, maxGSSize = 10000))

as.data.frame(enricher(filter(fittedBias, pvalue<.05 & ChrType=='autosomal' & log2FoldDiff<0)$Id, TERM2GENE=dplyr::select(gene_sets_filtered, set, ENSEMBL), pAdjustMethod='bonferroni', pvalueCutoff = .2, maxGSSize = 10000))


filter(gene_sets, set=='SandersASD') %>%
  left_join(dplyr::select(fittedBias, Id, padj), by=c("ENSEMBL" = "Id")) %>% View()


  dplyr::select(ENSEMBL) %>% 
  mutate(Background=1) %>% 
  right_join(filter(gene_sets_filtered, set=='SandersASD' & !is.na(ENSEMBL))) %>% 
  mutate(expressed=!is.na(Background)) %>%
  filter(!expressed)
  
filter(gene_sets_filtered, set=='SandersASD') %>% inner_join(DEall)
```

Correlation between counts and RIN
```{r}
cor <- left_join(DEgenes, counts) %>%
  dplyr::select(-SYMBOL, -Chr, -gene_type, -ChrType) %>%
  gather(label, norm_count, -Id) %>% 
  mutate(label=gsub('norm.', '', label )) %>%
  left_join(sample_info) %>% 
  group_by(Id) %>% 
  summarise(r=cor.test(norm_count, RIN)[[4]], p_val=cor.test(norm_count, RIN)[[3]])

cor_tr <- filter(fittedBias_tr, padj < 0.1) %>%
  dplyr::select(Id) %>%
  left_join(counts_tr) %>%
  dplyr::select(-SYMBOL, -Chr, -gene_type, -ChrType, -GeneId) %>%
  gather(label, norm_count, -Id) %>% 
  mutate(label=gsub('norm.', '', label )) %>%
  left_join(sample_info) %>%
  group_by(Id) %>%
  summarise(r=cor.test(norm_count, RIN)[[4]], p_val=cor.test(norm_count, RIN)[[3]])

cor_bg <- fittedBias %>% dplyr::select(Id) %>%
  left_join(counts) %>%
  dplyr::select(-SYMBOL, -Chr, -gene_type, -ChrType) %>%
  gather(label, norm_count, -Id) %>% 
  mutate(label=gsub('norm.', '', label )) %>%
  left_join(sample_info) %>%
  group_by(Id) %>%
  summarise(r=cor.test(norm_count, RIN)[[4]], p_val=cor.test(norm_count, RIN)[[3]])


filter(cor, p_val < 0.05) %>% nrow()/nrow(cor)
ggplot(cor, aes(x=r)) + geom_histogram() + main_theme()
ggplot(cor, aes(x=p_val)) + geom_histogram() + main_theme()

filter(cor_tr, p_val < 0.05) %>% nrow()/nrow(cor)
ggplot(cor_tr, aes(x=r)) + geom_histogram() + main_theme()
ggplot(cor_tr, aes(x=p_val)) + geom_histogram() + main_theme()

mutate(cor, DE='yes') %>%
  bind_rows(mutate(cor_bg, DE='no')) %>%
  ggplot(aes(x=r)) + geom_histogram() + facet_grid( DE ~ ., scales = "free") + main_theme()
```

Enrichment Analysis
Week 14
```{r}

#disease2gene <- filter(gene_sets, ! set %in% c('SandersASD', 'SFARI'))[, c("set", "gene_name")] 
disease2gene <- filter(gene_sets, ! set %in% c('SandersASD', 'SFARI')) %>% dplyr::select(set, gene_name) 
FemaleDEG <- fittedBias %>% filter(ageBin == '14' & padj<0.1 & log2FoldDiff <0)

FemaleEnrichment <- summary(enricher(FemaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
FemaleEnrichment$Direction <- 'FemaleBiased'

MaleDEG <- fittedBias %>% filter(ageBin == '14' & padj<0.1 & log2FoldDiff >0)

MaleEnrichment <- summary(enricher(MaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
MaleEnrichment$Direction <- 'MaleBiased'
Enrichment <- bind_rows(FemaleEnrichment, MaleEnrichment) %>% 
  separate(GeneRatio, c('x1', 'y1'), sep='/', remove=FALSE) %>% 
  separate(BgRatio, c('x2', 'y2'), sep='/', remove=FALSE) %>%
  mutate(Enrichment=(as.numeric(x1)/as.numeric(y1))/(as.numeric(x2)/as.numeric(y2))) %>%
  dplyr::select(Direction, ID, GeneRatio, BgRatio, Enrichment, pvalue, p.adjust, qvalue, geneID)

#Need to correct for 10 test because we tested 2 sets of DE genes
#For some mysterious reason, enricher is correcting for 4 tests even though there are clearly 5 gene sets being tested.
Enrichment$p.adjust <- p.adjust(Enrichment$pvalue, method='bonferroni', n=10)

write_tsv(Enrichment, "Tables/Enrichment14.txt")
Enrichment
```

## All weeks (65 high-cofindence Autism genes and 79 high confidence X-linked ID genes)
- Sanders et al 2015, Piton et al 2013
```{r}

disease2gene <- bind_rowsfilter(gene_sets, set %in% c('SandersASD', 'PitonID'))[, c("set", "gene_name")]

FemaleDEG <- fittedBiasAll %>% filter(padj<0.1 & log2FoldDiff < 0)

FemaleEnrichment <- summary(enricher(FemaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
dim(FemaleEnrichment)
#FemaleEnrichment$Direction <- 'FemaleBiased'

MaleDEG <- fittedBiasAll %>% filter(padj<0.1 & log2FoldDiff > 0)

MaleEnrichment <- summary(enricher(MaleDEG$SYMBOL, TERM2GENE=disease2gene, pAdjustMethod='bonferroni'))
MaleEnrichment$Direction <- 'MaleBiased'
Enrichment <- bind_rows(FemaleEnrichment, MaleEnrichment) %>% 
  separate(GeneRatio, c('x1', 'y1'), sep='/', remove=FALSE) %>% 
  separate(BgRatio, c('x2', 'y2'), sep='/', remove=FALSE) %>%
  mutate(Enrichment=(as.numeric(x1)/as.numeric(y1))/(as.numeric(x2)/as.numeric(y2))) %>%
  dplyr::select(Direction, ID, GeneRatio, BgRatio, Enrichment, pvalue, p.adjust, qvalue, geneID)

#Need to correct for test because 2 sets of DE genes
Enrichment$p.adjust <- p.adjust(Enrichment$pvalue, method='bonferroni', n=2)
write_tsv(Enrichment, "Tables/EnrichmentAll.txt")
Enrichment
```


Supplementary Table 1: Genes with significant sex Biases (FDR<0.1)
```{r}
DEG12_19 <- fittedBias %>% filter(padj<0.1 & !is.na(Chr)) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  mutate(Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl Gene ID` = Id, 
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG12_19), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS1',
            append=FALSE)
```

# Supplementary Table 2: Comparison with previous findings

```{r}
previous <- read_tsv("../Data/previous_findings.txt")
sig_res <- fittedBias %>%
  filter(padj < 0.1 & !is.na(Chr)) %>%
  mutate(result = paste0(format(log2FoldDiff, digits=2), ' (', format(padj, digits=2), ')'),
         Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) %>%
  dplyr::select(Id, SYMBOL, Chr, result) %>%
  arrange(Chr) %>% 
  dplyr::rename(`Ensembl ID` = Id, 
                `Gene ID` = SYMBOL,
                Chromosome = Chr,
                `Male Expression / Female (FDR-corrected p-value)` = result) %>%
  left_join(previous)

write.xlsx2(as.data.frame(sig_res), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS2',
            append=TRUE)
```

Supplementary Table 3: Transcripts with significant sex Biases (FDR<0.1)
```{r}
DEG12_19_tr <- fittedBias_tr %>% filter(padj<0.1 & !is.na(Chr)) %>%
  mutate(`FoldDiff` = 2^log2FoldDiff, 
         pvalue = as.numeric(format(pvalue, digits=2)), 
         padj = as.numeric(format(padj, digits=2))) %>%
  mutate(Chr = parse_factor(Chr, str_c('chr', c(seq(1,22), 'X', 'Y')))) %>%
  arrange(Chr) %>%
  dplyr::select(`Ensembl Transcript ID` = Id,
         `Ensembl Gene ID` = GeneId,
         `Gene ID` = SYMBOL,
         Chromosome = Chr,
         `Mean Count Male` = Male, 
         `Mean Count Female` = Female,
         `Fold Difference (M/F)` = FoldDiff,
         `p value` = pvalue,
         `FDR` = padj)

write.xlsx2(as.data.frame(DEG12_19_tr), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS3',
            append=TRUE)
```


Supplementary Table 4: Sampling info
```{r}
TotalFragments <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Tables/CountSummary.txt",
"\t", escape_double = FALSE, col_types = cols(Sample = col_character()),
trim_ws = TRUE) %>% dplyr::select(Sample, TotalFragments)

SampleInfo <- read_tsv("../Data/SampleInfo.txt", col_types = cols(Sample = col_character())) %>%
  full_join(TotalFragments) %>%
  mutate(`Total Read Pairs (millions)` = round(TotalFragments/1000000)) %>%
  dplyr::select(-TotalFragments)

write.xlsx2(as.data.frame(SampleInfo), 
            file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS4',
            append=TRUE)

```

Supplementary Table 5: X inactivation
```{r}
inner_join(XCI, fittedBias) %>% 
  dplyr::select(-Male, -Female, -Sex, -ChrType, -maxCooks) %>% 
  dplyr::select(Id, everything()) %>%
  mutate(Status=factor(Status, levels=c("E", "Mostly E", "VE",  "Mostly VE", "Mostly S", "S", "PAR", "No call", "Discordant" ))) %>%
  arrange(Status) %>%
  as.data.frame() %>% View()
  write.xlsx2(file="../Tables/SupplementalTables.xlsx", 
            row.names=FALSE,
            sheetName='TblS5',
            append=TRUE)

```

